pkg/proxy/iptables/proxier_test.go

```diff
@@ -694,6 +694,198 @@ func errorf(msg string, rules []iptablestest.Rule, t *testing.T) {
+			svc.Spec.ClusterIP = "10.20.30.43"
+			svc.Spec.Ports = []v1.ServicePort{{
+				Name:     "p80",
+				Port:     80,
+				Protocol: v1.ProtocolTCP,
+				NodePort: 3001,
+			}}
+		}),
+		// create ExternalIP service
+		makeTestService(makeNSN("ns4", "svc4").Namespace, "svc4", func(svc *v1.Service) {
+			svc.Spec.Type = "NodePort"
+			svc.Spec.ClusterIP = "10.20.30.44"
+			svc.Spec.ExternalIPs = []string{"50.60.70.81"}
+			svc.Spec.Ports = []v1.ServicePort{{
+				Name:       "p80",
+				Port:       80,
+				Protocol:   v1.ProtocolTCP,
+				TargetPort: intstr.FromInt(80),
+			}}
+		}),
+	)
+	makeEndpointsMap(fp,
+		// create ClusterIP service endpoints
+		makeTestEndpoints(makeNSN("ns1", "svc1").Namespace, "svc1", func(ept *v1.Endpoints) {
+			ept.Subsets = []v1.EndpointSubset{{
+				Addresses: []v1.EndpointAddress{{
+					IP: "10.180.0.1",
+				}},
+				Ports: []v1.EndpointPort{{
+					Name:     "p80",
+					Port:     80,
+					Protocol: v1.ProtocolTCP,
+				}},
+			}}
+		}),
+		// create LoadBalancer endpoints
+		makeTestEndpoints(makeNSN("ns2", "svc2").Namespace, "svc2", func(ept *v1.Endpoints) {
+			ept.Subsets = []v1.EndpointSubset{{
+				Addresses: []v1.EndpointAddress{{
+					IP: "10.180.0.2",
+				}},
+				Ports: []v1.EndpointPort{{
+					Name:     "p80",
+					Port:     80,
+					Protocol: v1.ProtocolTCP,
+				}},
+			}}
+		}),
+		// create NodePort service endpoints
+		makeTestEndpoints(makeNSN("ns3", "svc3").Namespace, "svc3", func(ept *v1.Endpoints) {
+			ept.Subsets = []v1.EndpointSubset{{
+				Addresses: []v1.EndpointAddress{{
+					IP: "10.180.0.3",
+				}},
+				Ports: []v1.EndpointPort{{
+					Name:     "p80",
+					Port:     80,
+					Protocol: v1.ProtocolTCP,
+				}},
+			}}
+		}),
+		// create ExternalIP service endpoints
+		makeTestEndpoints(makeNSN("ns4", "svc4").Namespace, "svc4", func(ept *v1.Endpoints) {
+			ept.Subsets = []v1.EndpointSubset{{
+				Addresses: []v1.EndpointAddress{{
+					IP:       "10.180.0.4",
+					NodeName: nil,
+				}, {
+					IP:       "10.180.0.5",
+					NodeName: utilpointer.StringPtr(testHostname),
+				}},
+				Ports: []v1.EndpointPort{{
+					Name:     "p80",
+					Port:     80,
+					Protocol: v1.ProtocolTCP,
+				}},
+			}}
+		}),
+	)
+
+	fp.syncProxyRules()
+
+	expected := `*filter
+:KUBE-SERVICES - [0:0]
+:KUBE-EXTERNAL-SERVICES - [0:0]
+:KUBE-FORWARD - [0:0]
+-A KUBE-FORWARD -m conntrack --ctstate INVALID -j DROP
+-A KUBE-FORWARD -m comment --comment "kubernetes forwarding rules" -m mark --mark 0x4000/0x4000 -j ACCEPT
+-A KUBE-FORWARD -m comment --comment "kubernetes forwarding conntrack pod source rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
+-A KUBE-FORWARD -m comment --comment "kubernetes forwarding conntrack pod destination rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
+COMMIT
+*nat
+:KUBE-SERVICES - [0:0]
+:KUBE-NODEPORTS - [0:0]
+:KUBE-POSTROUTING - [0:0]
+:KUBE-MARK-MASQ - [0:0]
+:KUBE-SVC-XPGD46QRK7WJZT7O - [0:0]
+:KUBE-SEP-SXIVWICOYRO3J4NJ - [0:0]
```

Comment:
hmm the test seems to rearrange the hash chains depending on when its created.

```
- [0:0]\n:KUBE-SVC-XPGD46QRK7WJZT7O 
- [0:0]\n:KUBE-SEP-SXIVWICOYRO3J4NJ 
- [0:0]\n:KUBE-SVC-GNZBNJ2PO5MGZ6GT 
- [0:0]\n:KUBE-XLB-GNZBNJ2PO5MGZ6GT 
- [0:0]\n:KUBE-FW-GNZBNJ2PO5MGZ6GT 
- [0:0]\n:KUBE-SEP-RS4RBKLTHTF2IUXJ 
- [0:0]\n:KUBE-SVC-X27LE4BHSL4DOUIK 
- [0:0]\n:KUBE-SEP-OYPFS5VJICHGATKP 
- [0:0]\n:KUBE-SVC-4SW47YFZTEDKD3PK 
- [0:0]\n:KUBE-SEP-UKSFD7AGPMPPLUHC 
- [0:0]\n:KUBE-SEP-C6EBXVWJJZMIWKLZ 
```
versus

```
- [0:0]\n:KUBE-SVC-GNZBNJ2PO5MGZ6GT 
- [0:0]\n:KUBE-XLB-GNZBNJ2PO5MGZ6GT
- [0:0]\n:KUBE-FW-GNZBNJ2PO5MGZ6GT 
- [0:0]\n:KUBE-SEP-RS4RBKLTHTF2IUXJ 
- [0:0]\n:KUBE-SVC-X27LE4BHSL4DOUIK 
- [0:0]\n:KUBE-SEP-OYPFS5VJICHGATKP 
- [0:0]\n:KUBE-SVC-4SW47YFZTEDKD3PK 
- [0:0]\n:KUBE-SEP-UKSFD7AGPMPPLUHC 
- [0:0]\n:KUBE-SEP-C6EBXVWJJZMIWKLZ 
- [0:0]\n:KUBE-SVC-XPGD46QRK7WJZT7O 
- [0:0]\n:KUBE-SEP-SXIVWICOYRO3J4NJ
```

I should probably sort it alphabetically to avoid this flake.

Reply:
Hi Surya,
Yes, this is a known issue in iptables-nft. User-defined chain ordering is random, sadly. This is on my TODO but I'm busy with other things at the moment and can't tell when it is going to be fixed.
Cheers, Phil

Reply:
> Hi Surya,
> Yes, this is a known issue in iptables-nft. User-defined chain ordering depends on order of creation, sadly. This is on my TODO but I'm busy with other things at the moment and can't tell when it is going to be fixed.
> Cheers, Phil

Hey Phil (didn't know n0-1 was you :) ). No issues I can hack this test.